### PR TITLE
fix: keep id if it exists

### PR DESCRIPTION
**Pull request by jobo322**
close: #99

### CODE REVIEW

**Comment by jobo322** [comment]
it is related with the discussion in https://github.com/cheminfo/nmr-processing/issues/205

**Comment by targos** [comment]
I tried a few ways to do it with a generic type but failed. Maybe it's possible but I don't know.

Here's the best I could come up with: https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABDAzgQQE4YIYE8DqMUAFgJIAmAFNlngFyJwBGAVgKbQDaAugJQM0cuZCkQBlKBhhgA5hR6IA3gChEaxBjZQQGJAHIY5PciSC8nAAzcA3MoC+y5aEiwEiYthQVKAN2wAbEDYGZnZofkQ-QLYRcUlpOXIlVXVNbV1EAyMTSICg2wdlKFwABxiJKVkKRABeJWRyBhR42UQHYrLEQhIKAB4AFQA+WsR+xAAyOMrEotKYgFk8JjZusnIB4bqxgB8uojWNx2koNgxgbAgYgGE4OAxyURV1RAAPBjAQAFtljFtn3HeXx+BUczmg8CQ5DgICY-jYNzuD18eWCiAR9xQEXRD1sYNckOhsPhtwxyOiDFWfWxKEGEUp62pg1x4HBbihMLh1LJQQYi1wy3pvUZET5Av2VJJD0GT3UMGAiEoHi8VCiQV4vGSz2eaR0SBlWq1hgYqrYADpDAAaFIG55vXLRU0vRAAKkQACYrTatQD7UFTcJXR7rVqHM87Ig2P4UDF9VqdRlY167SbHS73Z6veofSmA+ng2Hgw5Cn4MBo2CgAIwjdlErmKBoMPQnZp6C2vBgVts+t1tXi2TSV83kZQDiuO2zKEtllA9uo1zmSlCURR2zuIbt2Xgj8tuofbmeO5RAA

The problem with this implementation is that there is no check inside the implementation that the id is added in the right case (and having to define overloads for it is annoying).

/cc @stropitek Maybe you have an idea? Basically we want a function that accepts some type, possibly with an id inside. If there's an id, the return value contains an id as well, otherwise it doesn't. This playground show the idea (maybe it's a red herring and there's a completely different way to do it), but I don't know how to implement the body of the function: https://www.typescriptlang.org/play?#code/C4TwDgpgBAyg9gWwgFXNAvFA3gDwFxQB2ArggEYQBOAvgFCiRQDywAFlao5liASeVToAzYoQDGwAJZxCUYBADOwADzIoEHPMIATBbEQo0APgAUANwCGAG2IQCyAJT31miDr1Yok7QSWVJhADmUNRQAPzMbBxoUABk2N6+wP5BoQQs7JSc0Fi0UPlQlBDAxJSyuQWVULxQljYQAHQ4ADR5BdQA3LR0tJaUhYoAjFCY8komuASD1A60RQqDDSBzQw3etL0W-fMATCNyisAT+FCDzV4+UADkY8BXMysKO0uPz+tAA

**Comment by stropitek** [comment]
I've played around with the examples and I could not come up with a solution that circumvents the limitations @targos mentioned.

**Comment by jobo322** [comment]
I got a possible solution that avoid overloads, is it ok?

File: src/index.ts [hunk 1]
```diff
@@ -35,6 +36,10 @@ export interface OptimizedPeak {
   shape: Shape1D;
 }
 
+type PeakWithIDOrNot<T extends Peak> = T extends { id: string }
```
Comment (lpatiny) [inline src/index.ts:39]: Why do you extend `Peak` and not `OptimizedPeak`?

Also why call it `PeakWithIDOrNot` rather than `OptimizedPeakWithIDOrNot` ?

**Comment by lpatiny** [comment]
@targos This approach is what you expect ?

**Comment by targos** [comment]
@stropitek PTAL

File: src/index.ts (hunk 1 above)
Comment (jobo322) [reply src/index.ts:39]: There T is the type of input data, that is why I'm using `Peak`.

**Pull request merged**